src/main/scala/is/hail/annotations/ReadRowsRDD.scala

```diff
@@ -1,23 +1,200 @@
 package is.hail.annotations
 
-import java.io.{DataInputStream, DataOutputStream}
+import java.io.{InputStream, OutputStream}
+import java.util
 
 import is.hail.expr._
 import is.hail.utils.{SerializableHadoopConfiguration, _}
-import net.jpountz.lz4.{LZ4BlockInputStream, LZ4Factory}
+import is.hail.variant.LZ4Utils
 import org.apache.commons.lang3.StringUtils
 import org.apache.spark.rdd.RDD
-import org.apache.spark.sql.Row
 import org.apache.spark.{Partition, SparkContext, TaskContext}
 
-class Decoder(in: DataInputStream) {
+class ArrayInputStream(var a: Array[Byte], var end: Int) extends InputStream {
+  var off: Int = 0
 
-  def readByte(): Byte = {
-    val i = in.read()
-    assert(i != -1)
-    i.toByte
+  def this(a: Array[Byte]) = this(a, a.length)
+
+  def clear() {
+    off = 0
+  }
+
+  def read(): Int = {
+    if (off == end)
+      -1
+    else {
+      val c = a(off) & 0xff
+      off += 1
+      c
+    }
+  }
+
+  override def read(to: Array[Byte], toOff: Int, len: Int): Int = {
+    if (end == off)
+      -1
+    else {
+      val p = math.min(len, end - off)
+      System.arraycopy(a, off, to, toOff, p)
+      off += p
+      p
+    }
+  }
+
+  override def skip(n: Long): Long = {
+    assert(end != off)
+    assert(n <= Int.MaxValue)
+    val p = math.min(n.toInt, end - off)
+    off += p
+    p
+  }
+
+  override def available(): Int = end - off
+}
+
+class ArrayOutputStream(sizeHint: Int = 32) extends OutputStream {
+  var a: Array[Byte] = new Array[Byte](sizeHint)
+  var off: Int = 0
+
+  private def ensure(n: Int) {
+    val newEnd = off + n
+    if (newEnd > a.length) {
+      val newLength = math.max((a.length * 3) / 2, off + n)
```

Comment:
Why not just double?

Reply:
Same asymptotic complexity but this reduces memory  inefficiency.  I think I'll change it elsewhere, too, esp. in regions, which (eventually) will be the workhorse of memory usage.